OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,349 @@
+        }
+
+        public bool WasTapped(IModInputCombination combination)
+        {
+            return (!IsPressed(combination)) && (combination.PressDuration < TapDuration)
+                && combination.IsFirst;
+        }
+
+        public bool WasNewlyReleased(IModInputCombination combination)
+        {
+            return (!IsPressed(combination)) && combination.IsFirst;
+        }
+
+        private RegistrationCode SwapCombination(IModInputCombination combination, bool toUnregister)
+        {
+            if (combination.Hashes[0] <= 0)
+            {
+                return (RegistrationCode)combination.Hashes[0];
+            }
+            if (!toUnregister)
+            {
+                foreach (long hash in combination.Hashes)
+                {
+                    if (_comboRegistry.ContainsKey(hash) || (hash < MaxUsefulKey && _gameBindingCounter[hash] > 0))
+                    {
+                        return RegistrationCode.CombinationTaken;
+                    }
+                }
+            }
+            foreach (long hash in combination.Hashes)
+            {
+                if (toUnregister)
+                {
+                    _comboRegistry.Remove(hash);
+                    continue;
+                }
+                _comboRegistry.Add(hash, combination);
+            }
+            return RegistrationCode.AllNormal;
+        }
+
+        private List<string> GetCollisions(ReadOnlyCollection<long> hashes)
+        {
+            List<string> combos = new List<string>();
+            foreach (long hash in hashes)
+            {
+                if (_comboRegistry.ContainsKey(hash))
+                {
+                    combos.Add(_comboRegistry[hash].ModName + "." + _comboRegistry[hash].Name);
+                }
+                if (hash < MaxUsefulKey && _gameBindingCounter[hash] > 0)
+                {
+                    combos.Add("Outer Wilds." + Enum.GetName(typeof(KeyCode), (KeyCode)hash));
+                }
+            }
+            return combos;
+        }
+
+        public IModInputCombination RegisterCombination(IModBehaviour mod, string name, string combination)
+        {
+            var combo = new ModInputCombination(mod.ModHelper.Manifest, name, combination);
+            var code = SwapCombination(combo, false);
+            if (code == RegistrationCode.InvalidCombination)
+            {
+                _console.WriteLine("Failed to register \"" + mod.ModHelper.Manifest.Name + "." + name + "\": invalid combo!");
+            }
+            else if (code == RegistrationCode.CombinationTooLong)
+            {
+                _console.WriteLine("Failed to register \"" + mod.ModHelper.Manifest.Name + "." + name + "\": too long!");
+            }
+            else if (code == RegistrationCode.CombinationTaken)
+            {
+                _console.WriteLine("Failed to register \"" + mod.ModHelper.Manifest.Name + "." + name + "\": already in use by following mods:");
```

Comment:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated

Reply:
most probably done (also, I added FullName property to ModInputCombination to avoid that modName+"."+name thing)